File: vscode/src/lang/go/GoDepLensInlayProvider.ts
import * as vscode from "vscode";
import { Logger } from "../../utils/Logger";
import { GithubInlayUtils } from "../../utils/inlay/GithubInlayUtils";
import { GoDependencyParser } from "../../utils/parser/GoDependencyParser";
import { GithubRepoInfoService } from "../../utils/service/GithubRepoInfoService";
import { BaseDepLensInlayProvider } from "../BaseDepLensInlayProvider";

export class GoDepLensInlayProvider extends BaseDepLensInlayProvider {
  constructor(private readonly context: vscode.ExtensionContext) {
    super();

    GithubRepoInfoService.getInstance().onDidUpdateRepoInfo(() => {
      this.emitter.fire();
    });
  }

  protected isFileSupported(document: vscode.TextDocument): boolean {
    return document.languageId === "go" || document.fileName.endsWith("go.mod");
  }

  protected async provideInlayHintsForDocument(
    document: vscode.TextDocument,
    range: vscode.Range,
    token: vscode.CancellationToken,
  ): Promise<vscode.InlayHint[]> {
    const hints: vscode.InlayHint[] = [];
    try {
      const dependencies = await GoDependencyParser.parse(this.context, document, range);
      if (token.isCancellationRequested) {
        return [];
      }

      for (const dependency of dependencies) {
        GithubInlayUtils.addRepoInlay(
          hints,
          new vscode.Position(dependency.line, dependency.character),
          { owner: dependency.owner, repo: dependency.repo },
        );
      }
    } catch (error) {
      Logger.getInstance().error(`Failed to provide Go dependency inlay hints: ${String(error)}`);
      return [];
    }

    return hints;
  }
}

File: zed/lsp/src/lang/go/go_dep_lens_inlay_provider.rs
use crate::base_dep_lens_inlay_provider::BaseDepLensInlayProvider;
use crate::lsp::{InlayHint, Position, Range, TextDocument};
use crate::utils::inlay::github_inlay_utils::GithubInlayUtils;
use crate::utils::service::github_repo_info_service::GithubRepoInfoService;

pub struct GoDepLensInlayProvider;

impl GoDepLensInlayProvider {
    pub fn new() -> Self {
        GithubRepoInfoService::get_instance().on_did_update_repo_info_noop();
        Self
    }
}

impl BaseDepLensInlayProvider for GoDepLensInlayProvider {
    fn is_file_supported(&self, document: &TextDocument) -> bool {
        document.language_id == "go" || document.file_name.ends_with("go.mod")
    }

    fn provide_inlay_hints_for_document(
        &self,
        document: &TextDocument,
        range: &Range,
    ) -> Vec<InlayHint> {
        let mut hints = Vec::new();
        for dependency in dep_lens_lib::go::parse_go_dependencies(
            document.text(),
            &document.file_name,
            &document.language_id,
            range.start.line,
            range.end.line,
        ) {
            GithubInlayUtils::add_repo_inlay(
                &mut hints,
                Position {
                    line: dependency.line,
                    character: dependency.character,
                },
                dependency.owner,
                dependency.repo,
            );
        }

        hints
    }
}
